File: hyper_parallel/platform/mindspore/custom_ops/mhc_post.cc
/**
 * Copyright 2026 Huawei Technologies Co., Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <vector>
#include "ms_extension/all.h"
#include "module.h"

namespace custom {
namespace {
ms::Tensor GenResultTensor(const ms::Tensor &x) { return ms::Tensor(x.data_type(), x.shape()); }
}  // namespace

std::vector<ms::Tensor> npu_mhc_post(const ms::Tensor &x, const ms::Tensor &h_res, const ms::Tensor &h_out,
                                     const ms::Tensor &h_post) {
  auto out = GenResultTensor(x);
  ms::TensorToDevice(x, h_res, h_out, h_post);
  ms::TensorAllocate(out);
  MS_DISPATCH_ACLNN(aclnnMhcPost, x, h_res, h_out, h_post, out);
  return {out};
}

// cppcheck-suppress syntaxError
MS_CUSTOM_OPS_EXTENSION_MODULE(m) { m.def("npu_mhc_post", PYBOOST_CALLER(1, custom::npu_mhc_post)); }
}  // namespace custom

File: hyper_parallel/components/functional/mhc_post.py
# Copyright 2026 Huawei Technologies Co., Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ============================================================================
"""Manifold hyper-connection post-processing functions."""

from typing import Any, Tuple

import torch  # pylint: disable=forbidden-backend-import

try:
    import omni_training_custom_ops  # noqa: F401  # pylint: disable=unused-import
except ImportError:
    omni_training_custom_ops = None


class _MhcPost(torch.autograd.Function):
    """Autograd bridge for the NPU MHC post custom operator."""

    @staticmethod
    def forward(
        ctx: Any,
        x: torch.Tensor,
        h_res: torch.Tensor,
        h_out: torch.Tensor,
        h_post: torch.Tensor,
    ) -> torch.Tensor:
        """Run the NPU MHC post forward operator."""
        if omni_training_custom_ops is None:
            raise ImportError("MHC AscendC requires omni_training_custom_ops")
        ctx.save_for_backward(x, h_res, h_out, h_post)
        return torch.ops.custom.npu_ai_infra_manifold_constrained_hyper_connection_post(
            x,
            h_res,
            h_out,
            h_post,
        )

    @staticmethod
    def backward(
        ctx: Any,
        grad_output: torch.Tensor,
    ) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
        """Run the NPU MHC post backward operator."""
        x, h_res, h_out, h_post = ctx.saved_tensors
        grad_x, grad_h_res, grad_h_out, grad_h_post = torch.ops.custom.npu_ai_infra_mhc_post_grad(
            grad_output.contiguous(),
            x,
            h_res,
            h_out,
            h_post,
        )
        return grad_x, grad_h_res, grad_h_out, grad_h_post


def mhc_post(
    x: torch.Tensor,
    residual: torch.Tensor,
    h_post: torch.Tensor,
    h_res: torch.Tensor,
    num_stream: int,
) -> torch.Tensor:
    """Combine transformed and residual streams after an MHC-wrapped block.

    Args:
        x: Output of the wrapped block.
        residual: Flattened residual streams.
        h_post: Per-stream output mixing coefficients.
        h_res: Residual stream mixing matrix.
        num_stream: Number of residual streams.

    Returns:
        Flattened mixed residual streams.
    """
    x_shape = x.size()
    residual = residual.reshape(x_shape[0], x_shape[1], num_stream, -1)
    return _MhcPost.apply(residual, h_res, x, h_post).flatten(2)
